File: Solutions/javascript/2022/day_04.js
const { getPuzzleInput } = require("../SetupSolutions/SetupSolutions.js");

const year = "2022";
const day = "04";
const puzzleInput = getPuzzleInput(year, `Day_${day}.txt`).filter(el => el);
const puzzle = puzzleInput;

const getNumOfAllFullyContainedPairs = (puzzle) => {
    let total = 0;
    for (const line of puzzle) {
        const [one, two] = line.split(",");
        const [min_one, max_one] = one.split("-").map(el =>parseInt(el));
        const [min_two, max_two] = two.split("-").map(el =>parseInt(el));

        if ((min_one >= min_two && max_one <= max_two) || 
        (min_two >= min_one && max_two <= max_one)) {
            total += 1;
        }
    }

    return total;
};

const getNumOfAllOverlappingPairs = (puzzle) => {
    let total = 0;
    for (const line of puzzle) {
        const [one, two] = line.split(",");
        const [min_one, max_one] = one.split("-").map(el =>parseInt(el));
        const [min_two, max_two] = two.split("-").map(el =>parseInt(el));
        let list1 = []
        let list2 = []
        for (let i = min_one; i <= max_one; i++) {
            list1.push(i);
        }
        for (let i = min_two; i <= max_two; i++) {
            list2.push(i);
        }

        if (list2.some(el => el === min_one || el === max_one) 
        || list1.some(el => el === min_two || el === max_two)) {
            total += 1;
        }
    }

    return total;
};

// Script runs here
console.log(`Advent of Code - ${year}: Day ${day}`);
console.log("Part-One Answer:", getNumOfAllFullyContainedPairs(puzzle));
console.log("Part-Two Answer:", getNumOfAllOverlappingPairs(puzzle));

File: Solutions/python/2022/day_04.py
import sys, os
sys.path.append(os.path.abspath(os.path.join(os.path.realpath(os.path.dirname(__file__)), "..", "SetupSolutions")))

from SetupSolutions import getPuzzleInput

year = os.path.basename(os.path.dirname(__file__))
day = os.path.basename(__file__).split(".")[0].capitalize()
puzzleInput = [line.replace("\n", "") for line in getPuzzleInput(f"{year}", f"{day}.txt")]
puzzle = puzzleInput

def findFullyContainedPairs(puzzle):
    fullyContainedPairs = 0

    for line in puzzle:
        assignments = line.replace("\n", "").split(",")
        first = assignments[0].split("-")
        second = assignments[1].split("-")

        min_first = int(first[0])
        max_first = int(first[1])
        min_second = int(second[0])
        max_second = int(second[1])

        if (min_first >= min_second and max_first <= max_second) \
            or (min_second >= min_first and max_second <= max_first):
            fullyContainedPairs += 1
    
    return fullyContainedPairs

def findOverlappingPairs(puzzle):
    overlappingPairs = 0

    for line in puzzle:
        assignments = line.replace("\n", "").split(",")
        first = assignments[0].split("-")
        second = assignments[1].split("-")

        min_first = int(first[0])
        max_first = int(first[1])
        min_second = int(second[0])
        max_second = int(second[1])

        range_1 = [i for i in range(min_first, max_first + 1)]
        range_2 = [i for i in range(min_second, max_second + 1)]

        if (min_first in range_2 or max_first in range_2) \
            or (min_second in range_1 or max_second in range_1):
            overlappingPairs += 1
    
    return overlappingPairs

# Script runs here
print(f"Advent of Code - {year}: {day.replace('_', ' ')}")
print("Part-One Answer:", findFullyContainedPairs(puzzle))
print("Part-Two Answer:", findOverlappingPairs(puzzle))